### DISCUSSION TITLE
LoraWan Connection issue with ABP

### DISCUSSION THREAD

**Discussion by Ahsan-Averos** [Lora-net/LoRaMac-node#1093] [Q&A]
Dear Sir
Please correct me what I am doing wrong, I done it number of serval trial to connect with gateway but unable to do this.
Although my Lora communication is working fine.

Reference code:
[(https://github.com/insightsip/ISP4520-examples)]

**Hardware:**

nrf52832
E22900M22S(Sx1262)

**Please see the response in log file.**

_> app: LoRaWan end-device started.
<info> app: Initializing fds...
<info> app: Reading flash usage statistics...
<info> app: Found 7 valid records.
<info> app: Found 75 dirty records (ready to be garbage collected).
<debug> app: Record 3 at 7D06C
<debug> app: Record 4 at 7D1D0
<debug> app: Record 6 at 7D41C
<debug> app: Record 7 at 7D4F0
<debug> app: Record 80 at 7ED9C
<debug> app: Record 81 at 7EDDC
<debug> app: Record 82 at 7EE00
<info> app: Data restored from NVM: 1208 bytes
<info> app: MLME-Request: MLME_JOIN, OK
<debug> app: Uplink counter   : 26
<debug> app: Uplink Data rate : DR_0
<debug> app: Uplink TX power  : 0
<debug> app: Uplink frequency: 868500000
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x1
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x2
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x5
<info> app: Data stored on NVM: 224 bytes
<info> app: MCPS-Request: MCPS_UNCONFIRMED,  OK
<debug> app: Uplink counter   : 27
<debug> app: Uplink Data rate : DR_0
<debug> app: Uplink TX power  : 0
<debug> app: Uplink frequency: 868300000
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x1
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x2
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x5
<info> app: Data stored on NVM: 224 bytes
<info> app: MCPS-Request: MCPS_UNCONFIRMED,  OK
<debug> app: Uplink counter   : 28
<debug> app: Uplink Data rate : DR_0
<debug> app: Uplink TX power  : 0
<debug> app: Uplink frequency: 868300000
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x1
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x2
<debug> app: FDS update success! FileId: 0x3E8 RecKey:0x5
<info> app: Data stored on NVM: 224 bytes
<info> app: MCPS-Request: MCPS_UNCONFIRMED,  OK_

**And also see what I changed in code** 

 
![image](https://user-images.githubusercontent.com/81947656/117609652-d3f2dc80-b179-11eb-9ac4-286bcb7b7387.png)
![image](https://user-images.githubusercontent.com/81947656/117610487-46b08780-b17b-11eb-97f0-8f3b482f3b94.png)
![image](https://user-images.githubusercontent.com/81947656/117609736-f4229b80-b179-11eb-9314-d8a909052667.png)

**Server data**
Device address:

![image](https://user-images.githubusercontent.com/81947656/117610056-89be2b00-b17a-11eb-8d8c-b4f5bd765f0c.png)

Network + App  session keys:

![image](https://user-images.githubusercontent.com/81947656/117610110-a195af00-b17a-11eb-8452-a47f0d9228f6.png)

Graph:

![image](https://user-images.githubusercontent.com/81947656/117610165-b7a36f80-b17a-11eb-976e-246dc5035686.png)

I hope that someone helping me out in this.
Thanks in advance.....

**Comment by mluis1** [comment]
First I would like to clarify something. An end-device does not connect to a gateway. An end-device connects to a network server.
A gateway is just a dumb device that translates what is received on the RF to the IP world and vice versa.

Which Network Server are you using?

Most of the network servers that I'm aware of use big endian encoding for the `DevAddr`. The LoRaMac-node project also encodes the `DevAddr` as big endian (see comment above `LORAWAN_DEVICE_ADDRESS` definition).

On your screen shot you have entered the `DevAddr` encoded as little endian.

`DevAddr` should be: `60595857`

In the future it would be nice if you could post this kind of questions on the project [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab. It is a better place to engage discussions and then we can agree if it is an issue or not.
If you don't mind I move this issue to the [Discussions](https://github.com/Lora-net/LoRaMac-node/discussions) tab.

**Reply by Ahsan-Averos** [reply]
Dear Sir,
Yes I know, my end device it connected with personalize server which is inspired by https://github.com/gotthardp/lorawan-server, you can check. 

_On your screen shot you have entered the DevAddr encoded as little endian._
I do not understand in source code I mention 60595857, so in this case big endian should be 57585960 on server configuration, am I right? and what about session keys?

**Reply by Ahsan-Averos** [reply]
If it is possible please send me guide how can I set end device to ABP, what I understand my setting is fine in both files:
Commissioning.h and se-identity.h
But still I am not receiving any data?

**Reply by mluis1** [reply]
On source code you read the value in big endian. It is just over the air that the `DevAddr` is encoded in little endian (protocol specification). The MAC layer takes care of this conversion.
You just need to copy it as is to your network server.

The keys look correct. (I did not check nibble by nibble)
In addition a key has no endianess it is just a random suite of bytes. So, one just copy it as is.

**Reply by Ahsan-Averos** [reply]
Waiting for your reply!

**Reply by Ahsan-Averos** [reply]
Sir for ABP:

Network session key is same in this all byte:

F_NWK_S_INT_KEY
S_NWK_S_INT_KEY
NWK_S_ENC_KEY

and application session key.

APP_S_KEY
 is this okay?

**Reply by Ahsan-Averos** [reply]
1B2E354658AED2A6ABF7158809CF4F3C (Network session key)
5B4E352618AED2A6ABF7158809CF4F3C (Apps session key).

**Reply by Ahsan-Averos** [reply]
Now receiving this :
![image](https://user-images.githubusercontent.com/81947656/117623275-ffcb8d80-b18c-11eb-8b3f-ec95dba67184.png)

**Reply by Ahsan-Averos** [reply]
after change to DevAddr to Big  Endian 60595857

**Reply by mluis1** [reply]
Personally, I used that network server very long time ago (just a couple of hours trial) and I don't remember how the end-device must be setup on it.

From the screen shots you show the only thing that look to be wrong is the `DevAddr` value. However, I don't know what is expected by the network server you are using.

Maybe, @altishchenko can help you on this.

As shown in comments above keys definition the `F_NWK_S_INT_KEY` = `S_NWK_S_INT_KEY` = `NWK_S_ENC_KEY` = `NWK_S_KEY` in case of LoRaWAN 1.0.x specifications.

Key definitions:
```c
        {                                                                                                           \
            /*!                                                                                                     \
             * Forwarding Network session integrity key                                                             \
             * WARNING: NWK_S_KEY FOR 1.0.x DEVICES                                                                 \
             */                                                                                                     \
            .KeyID    = F_NWK_S_INT_KEY,                                                                            \
            .KeyValue = { 0x2B, 0x7E, 0x15, 0x16, 0x28, 0xAE, 0xD2, 0xA6, 0xAB, 0xF7, 0x15, 0x88, 0x09, 0xCF, 0x4F, \
                          0x3C },                                                                                   \
        },                                                                                                          \
        {                                                                                                           \
            /*!                                                                                                     \
             * Serving Network session integrity key                                                                \
             * WARNING: NOT USED FOR 1.0.x DEVICES. MUST BE THE SAME AS \ref LORAWAN_F_NWK_S_INT_KEY                \
             */                                                                                                     \
            .KeyID    = S_NWK_S_INT_KEY,                                                                            \
            .KeyValue = { 0x2B, 0x7E, 0x15, 0x16, 0x28, 0xAE, 0xD2, 0xA6, 0xAB, 0xF7, 0x15, 0x88, 0x09, 0xCF, 0x4F, \
                          0x3C },                                                                                   \
        },                                                                                                          \
        {                                                                                                           \
            /*!                                                                                                     \
             * Network session encryption key                                                                       \
             * WARNING: NOT USED FOR 1.0.x DEVICES. MUST BE THE SAME AS \ref LORAWAN_F_NWK_S_INT_KEY                \
             */                                                                                                     \
            .KeyID    = NWK_S_ENC_KEY,                                                                              \
            .KeyValue = { 0x2B, 0x7E, 0x15, 0x16, 0x28, 0xAE, 0xD2, 0xA6, 0xAB, 0xF7, 0x15, 0x88, 0x09, 0xCF, 0x4F, \
                          0x3C },                                                                                   \
        },                                                                                                          \
        {                                                                                                           \
            /*!                                                                                                     \
             * Application session key                                                                              \
             */                                                                                                     \
            .KeyID    = APP_S_KEY,                                                                                  \
            .KeyValue = { 0x2B, 0x7E, 0x15, 0x16, 0x28, 0xAE, 0xD2, 0xA6, 0xAB, 0xF7, 0x15, 0x88, 0x09, 0xCF, 0x4F, \
                          0x3C },                                                                                   \
        },                                                                                                          \
```

Something you don't say is if the uplinks are being received by the packet_forwarder on the gateway and if they are forwarded to the network server.
Please try to look at the gateway logs and network server logs to see if the uplinks are being received.

**Comment by altishchenko** [comment]
Hello, as I can see, this is Petr's lorawan server.
1. Miguel is right about your DevAddr - it should be the same in both places.
2. Also, make sure that 'FCnt Check' in your device's profile is set to 'Reset on zero', you are using ABP and will definitely reset your device multiple times, this setting will help.
3. In your 'Areas' settings set 'Log Ignored' to 'true', this allows you to see the fact of device transmission even if it doesn't match your configured devaddr or keys.
4. Check your server's debug.log.
5. You are getting *ack_lost* messages, this means that your server to gateway connection is not very good or there is some misconfiguration on the TCP/IP level between the two.

**Reply by Ahsan-Averos** [reply]
Dear @altishchenko 
Thanks for your reply.
2.where to reset frame counter as I already reset is to 0.

![image](https://user-images.githubusercontent.com/81947656/117625782-d102e680-b18f-11eb-9262-0fa324163423.png)

3. 
![image](https://user-images.githubusercontent.com/81947656/117626101-2fc86000-b190-11eb-8b34-bee40b6bf1b4.png)

**Reply by Ahsan-Averos** [reply]
I already done this thing.
Although Sir when used this code with my board is working fine below is link:

https://github.com/beegee-tokyo/SX126x-Arduino/tree/master/examples/LoRaWan
But I do not want to use in arduino.

**Reply by altishchenko** [reply]
@Ahsan-Averos did you configure any Networks, Groups, Areas or Profiles for your installation? I am quite sure, that without proper config things just won't work as expected.

**Reply by Ahsan-Averos** [reply]
Dear Sir,
Can you please guide me how can I configure my setting for **_EU868_**? I am also send you some picture please correct me.
Thanks

**Reply by Ahsan-Averos** [reply]
**Network List::**

![image](https://user-images.githubusercontent.com/81947656/117762692-12a09980-b243-11eb-9f18-2d4bcbcbfec6.png)

**General Setting::**

![image](https://user-images.githubusercontent.com/81947656/117762751-26e49680-b243-11eb-915c-ac7f6c8114ab.png)

**ADR Setting::**

![image](https://user-images.githubusercontent.com/81947656/117762823-467bbf00-b243-11eb-8f73-69078d7637f6.png)

**Channel Setting::**

![image](https://user-images.githubusercontent.com/81947656/117762847-54c9db00-b243-11eb-8c9b-7c133af3a3cb.png)

**Group Setting::**

![image](https://user-images.githubusercontent.com/81947656/117762967-762ac700-b243-11eb-94d9-3c4a3f2a3bf2.png)

**Profile General::**

![image](https://user-images.githubusercontent.com/81947656/117763090-a7a39280-b243-11eb-8817-bc2d0c46bd10.png)

**Profile ADR::**

![image](https://user-images.githubusercontent.com/81947656/117763119-b38f5480-b243-11eb-971d-4f5cab16e2f5.png)

**Node General::**

![image](https://user-images.githubusercontent.com/81947656/117763253-f05b4b80-b243-11eb-9222-0d9ce9cf105e.png)

**Node ADR::**

![image](https://user-images.githubusercontent.com/81947656/117763298-0406b200-b244-11eb-8dc9-024ae7d06130.png)

**Ignored Node list::**

![image](https://user-images.githubusercontent.com/81947656/117763361-1bde3600-b244-11eb-8c83-c8cad229da89.png)

I am waiting for your reply

**Reply by Ahsan-Averos** [reply]
Please see the above setting and let me know the parameter according to code, may be some delay, power or other issue.
Thanks

**Reply by altishchenko** [reply]
Great! Give me some time please and I'll get back to you as soon as possible!

**Reply by Ahsan-Averos** [reply]
Okay Sir, I will waiting for your reply!
But I am still curious why is working with Arduino code. That I mention above.
Thanks

**Reply by altishchenko** [reply]
Ok. I am back. Let's start from the top, step by step.

Area - something global, Gateways belong to Areas, Areas define the radio regional settings for the gateways, your Area should like this (mind the red selection!):
![p1](https://user-images.githubusercontent.com/19464159/117806646-44c1f380-b263-11eb-8b22-7802b2641785.PNG)

**Reply by altishchenko** [reply]
Gateways, simple devices. This is an example of my Kerlink Wirnet station:
![p2](https://user-images.githubusercontent.com/19464159/117807002-b9952d80-b263-11eb-9857-6499a7b59c6a.PNG)

**Reply by altishchenko** [reply]
Networks - define radio regional settings for **devices**, this is a working example for EU868. Pay attention to **commas** in the screens below. For the beginning, make sure your settings match mine!

General
![p3](https://user-images.githubusercontent.com/19464159/117807879-e138c580-b264-11eb-80b1-9ee4f28a6b1b.PNG)

ADR
![p4](https://user-images.githubusercontent.com/19464159/117807912-ec8bf100-b264-11eb-82db-ee50859cc658.PNG)

Channels
![p5](https://user-images.githubusercontent.com/19464159/117807934-f7df1c80-b264-11eb-9b7a-67a322e00b7b.PNG)

**Reply by altishchenko** [reply]
Groups - define sets of devices belonging to a particular network and persons who can manage them:

![p6](https://user-images.githubusercontent.com/19464159/117808390-7d62cc80-b265-11eb-9edf-bf6b9bb5322f.PNG)

**Reply by altishchenko** [reply]
Your profile and node configuration look pretty decent to me, with one correction - I'd leave the **TX Window** setting in your profile to **Auto**, this way, if anything goes wrong, the server will be able to retry ACK transmission in the standard RX2 window.

**Comment by Ahsan-Averos** [comment]
Hello Sir,
Sorry for late reply, I done all the things that you told me but still is not working!!!

**Reply by altishchenko** [reply]
Ok, @Ahsan-Averos can you send me your network server debug.log file?
